File: library/src/main/java/com/yzyfdf/library/base/BaseRvAdapter.java
package com.yzyfdf.library.base;

import android.animation.AnimatorSet;
import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;

import java.util.ArrayList;
import java.util.List;


public abstract class BaseRvAdapter<VH extends RecyclerView.ViewHolder, T> extends RecyclerView.Adapter<VH> {

    public Context mContext;
    public LayoutInflater mInflater;
    public List<T> mList;
    public OnItemClickListener mOnItemClickListener;
    public OnNoItemCallback mOnNoItemCallback;

    private boolean showEnterAnim;
    private int mLastPosition = -1;

    public BaseRvAdapter(Context context) {
        this(context, new ArrayList<T>());
    }

    public BaseRvAdapter(Context context, List<T> list) {
        mContext = context;
        mList = list;
        mInflater = LayoutInflater.from(context);
    }

    @Override
    public int getItemCount() {
        return mList == null ? 0 : mList.size();
    }

    @Override
    public void onBindViewHolder(VH holder, int position) {

        //第一次显示动画
        if (showEnterAnim && position > mLastPosition) {
            AnimatorSet set = getAnim(holder);
            set.start();
            mLastPosition = position;
        }

        //onBindViewHolder
        onBindVH(holder, position);
    }

    /**
     * onBindViewHolder
     */
    public abstract void onBindVH(VH holder, int position);

    /**
     * 是否显示条目进入动画
     */
    public void setShowEnterAnim(boolean showEnterAnim) {
        this.showEnterAnim = showEnterAnim;
    }

    /**
     * 换成自己想要的动画
     */
    public AnimatorSet getAnim(VH holder) {
        return RvAnimator.getInstance().getAnim(holder.itemView);
    }

    /**
     * @param list 刷新
     */
    public void refresh(List<? extends T> list) {
        if (list == null) {
            return;
        }
        refresh(list, true);
    }

    /**
     * @param list 刷新
     * @param force 删除原来数据或者增加在最前面
     */
    public void refresh(List<? extends T> list, boolean force) {
        if (list == null) {
            return;
        }
        if (force) {
            //强制刷新 清楚原来所有数据
            mList.clear();
            mList.addAll(list);
        } else {
            mList.addAll(0, list);
        }
        notifyDataSetChanged();
    }

    /**
     * @param list 增加集合
     */
    public void add(List<? extends T> list) {
        if (list == null) {
            return;
        }
        mList.addAll(list);
        notifyDataSetChanged();
    }

    /**
     * @param t 增加
     */
    public void add(T t) {
        insert(t, mList.size());
    }

    /**
     * @param t 对象
     * @param i 位置
     */
    public void insert(T t, int i) {
        if (t == null) {
            return;
        }
        if (i > mList.size()) {
            i = mList.size();
        }
        mList.add(i, t);
        notifyDataSetChanged();
    }

    public void clear() {
        mList.clear();
        notifyDataSetChanged();
    }

    /**
     * @param index 索引
     */
    public void remove(int index) {
        mList.remove(index);
        notifyItemRemoved(index);
        notifyItemRangeChanged(index, mList.size() - index);
    }


    /**
     * @param t 对象
     */
    public void remove(T t) {
        if (t == null) {
            return;
        }
        for (int i = 0; i < mList.size(); i++) {
            if (mList.get(i) == t) {
                remove(i);
                return;
            }
        }
    }

    public List<T> getList() {
        return mList;
    }


    /**
     * @param <T> 条目被点击
     */
    public interface OnItemClickListener<T> {
        void onItemClick(T t, int i);
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        mOnItemClickListener = onItemClickListener;
    }

    /**
     * 全部删除，没有条目了
     */
    public interface OnNoItemCallback {
        void noItem();
    }

    public void setOnNoItemCallback(OnNoItemCallback onNoItemCallback) {
        mOnNoItemCallback = onNoItemCallback;
    }


}

File: library/src/main/java/com/yzyfdf/library/base/BaseRvAdapter.kt
package com.yzyfdf.library.base

import android.animation.Animator
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import java.util.*
import kotlin.math.min


abstract class BaseRvAdapter<VH : RecyclerView.ViewHolder, T>
@JvmOverloads constructor(var mContext: Context, var mList: MutableList<T> = ArrayList())
    : RecyclerView.Adapter<VH>() {

    var mInflater: LayoutInflater = LayoutInflater.from(mContext)
    lateinit var mOnItemClickListener: (adapter: BaseRvAdapter<VH, T>, view: View, position: Int, item: T) -> Unit
    lateinit var mOnNoItemCallback: () -> Unit

    private var showEnterAnim: Boolean = false
    private var animator: RvAnimator? = null

    private var mLastPosition = -1

    val list: List<T>
        get() = mList

    override fun getItemCount(): Int {
        return mList.size
    }

    override fun onBindViewHolder(holder: VH, position: Int) {

        //第一次显示动画
        if (showEnterAnim && position > mLastPosition) {
            getAnim(holder)?.forEach {
                it.start()
            }
            mLastPosition = position
        }

        //onBindViewHolder
        onBindViewHolder(holder, position, mList[position])

        if (this::mOnItemClickListener.isInitialized) {
            holder.itemView.setOnClickListener { view ->
                mOnItemClickListener(this, holder.itemView, position, mList[position])
            }
        }

    }

    /**
     * onBindViewHolder
     */
    abstract fun onBindViewHolder(holder: VH, position: Int, itemBean: T)

    /**
     * 是否显示条目进入动画
     */
    fun setShowEnterAnim(showEnterAnim: Boolean, animator: RvAnimator = RvAnimator.SlideInRightAnimation()) {
        this.showEnterAnim = showEnterAnim
        this.animator = animator
    }

    /**
     * 换成自己想要的动画
     */
    fun getAnim(holder: VH): Array<Animator>? {
        return animator?.getAnims(holder.itemView)
    }

    /**
     * @param list 刷新
     */
    fun refresh(list: List<T>?) {
        list?.let {
            refresh(it, true)
        }
    }

    /**
     * @param list  刷新
     * @param force 删除原来数据或者增加在最前面
     */
    fun refresh(list: List<T>?, force: Boolean) {
        list?.let {
            if (force) {
                //强制刷新 清楚原来所有数据
                mList.clear()
                mList.addAll(it)
            } else {
                mList.addAll(0, it)
            }
            notifyDataSetChanged()

        }
    }

    /**
     * @param list 增加集合
     */
    fun add(list: List<T>?) {
        list?.let {
            mList.addAll(it)
            notifyDataSetChanged()
        }
    }

    /**
     * @param t 增加
     */
    fun add(t: T) {
        insert(t, mList.size)
    }

    /**
     * @param t 对象
     * @param i 位置
     */
    fun insert(t: T?, i: Int) {
        t?.let {
            mList.add(min(i, mList.size), it)
            notifyDataSetChanged()
        }
    }

    fun clear() {
        mList.clear()
        notifyDataSetChanged()
    }

    /**
     * @param index 索引
     */
    fun remove(index: Int) {
        if (index >= 0 && index < mList.size) {
            mList.removeAt(index)
            notifyItemRemoved(index)
            notifyItemRangeChanged(index, mList.size - index)
        }
    }


    /**
     * @param t 对象
     */
    fun remove(t: T?) {
        t?.let {
            for (i in mList.indices) {
                if (it == mList[i]) {
                    remove(i)
                    return
                }
            }
        }
    }

    fun setOnItemClickListener(listener: (adapter: BaseRvAdapter<VH, T>, view: View, position: Int, item: T) -> Unit) {
        mOnItemClickListener = listener
    }

    fun setOnNoItemCallback(listener: () -> Unit) {
        mOnNoItemCallback = listener
    }


}
